sql/src/SqlCsv.sk

```diff
@@ -728,75 +688,95 @@ fun replayDiff(
         tableDiffs.push(currentDiff);
         checkpoint = parseCheckpoint(line);
-        break SKStore.CContinue(
-          buildDiffChain(tableDiffs, identity, userFileOpt, checkpoint),
-        )
-
-      | current @ DPBuildingTableDiff(_, currentDiff) if (line == "\t\t") ->
-        currentDiff.!reset = true;
-        current
-
-      | current @ DPBuildingTableDiff(tableDiffs, currentDiff) ->
-        currentDiff.lines.push(line); // capture the raw input for if we end up rejecting the txn
 
-        parseTableChange(userFileOpt, context, currentDiff.table, line) match {
-        | Some(rows) ->
-          currentDiff.rows.extend(rows);
-          current
-        | None() ->
+        rejectedAnyDiff = false;
+        userFileOpt.each(userFile ->
+          SKStore.withContext(privacyContext -> {
+            accessSolver = SKDB.AccessSolver::create(
+              privacyContext,
+              userFile.name,
+            );
+            for (diff in tableDiffs) {
+              if (rejectedAnyDiff) {
+                break void
+              };
+              tableName = diff.table.name.origName;
+              authorCol = SKDB.getAuthorIdx(diff.table);
+              accessCol = SKDB.getAccessIdx(diff.table);
+              for (row in diff.rows) {
+                SKDB.checkUserCanWriteRow(
+                  privacyContext,
+                  accessSolver,
+                  tableName,
+                  authorCol,
+                  accessCol,
+                  row,
+                ) match {
+                | SKDB.AROK() -> void
+                | SKDB.ARError(err) ->
+                  print_error("Error: " + err.toString());
+                  !rejectedAnyDiff = true;
+                  break void
+                }
+              };
+            };
+          })
+        );
+        if (rejectedAnyDiff) {
           writer = mutable Debug.BufferedWriter(print_raw, 4096);
-          tableDiffs.push(currentDiff);
-          for (diff in tableDiffs) {
-            nack(writer, diff)
+          for (tdiff in tableDiffs) {
+            nack(writer, tdiff)
           };
-          mutable DPRejectingTxn(writer)
+          writer.flush();
+          break SKStore.CContinue(None())
+        } else {
+          break SKStore.CContinue(
+            buildDiffChain(tableDiffs, identity, userFileOpt, checkpoint),
+          )
         }
 
-      | current @ DPRejectingTxn(writer) if (line.startsWith("^")) ->
-        (table, snapshot) = parseHeader(context, line);
-        nack(writer, TableDiff::create(table, snapshot));
+      | current @ Some(DiffParsingState(_, currentDiff)) if (line == "\t\t") ->
+        currentDiff.!reset = true;
         current
 
-      | DPRejectingTxn(writer) if (line.startsWith(":")) ->
-        writer.write(line);
-        writer.write("\n");
-        writer.flush();
-        break SKStore.CContinue(None())
-
-      | current @ DPRejectingTxn(writer) ->
-        writer.write(line);
```

Comment:
[no action] It's a shame we lost the rejecting txn state as this saved buffering the entire input. But I can live with it. We should probably protect ourselves from buffering too much here regardless.